src/gpu/intel/logging.hpp

```diff
@@ -0,0 +1,173 @@
+/*******************************************************************************
+* Copyright 2025 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#ifndef GPU_INTEL_LOGGING_HPP
+#define GPU_INTEL_LOGGING_HPP
+
+#include <iostream>
+#include <sstream>
+#include <string>
+
+#include "gpu/intel/utils.hpp"
+
+namespace dnnl {
+namespace impl {
+namespace gpu {
+namespace intel {
+
+enum class log_level_t {
+    off = 0,
+    warning = 100,
+    suggestion = 120,
+    info = 150,
+    perf = 170,
+    trace = 200,
+};
+
+template <typename T, typename = void>
+struct print_helper_t {
+    static void call(std::ostream &out, const T &t) { out << t; }
+};
+
+template <typename T>
+struct print_helper_t<T, decltype(std::declval<T>().str(), void())> {
+    static void call(std::ostream &out, const T &t) { out << t.str(); }
+};
+
+template <log_level_t level, bool value = true>
+class logger_t {
+public:
+    logger_t(const char *file_name, int line, std::ostream &out = std::cout)
+        : file_path_(file_name + std::string(":") + std::to_string(line))
+        , out_(out) {
+        auto pos = std::min(file_path_.rfind("\\"), file_path_.rfind("/"));
+        if (pos != std::string::npos) file_path_ = file_path_.substr(pos + 1);
+    }
+    ~logger_t() {
+        if (printed_ && !ends_with_endl_) out_ << std::endl;
+    }
+
+    static bool is_enabled() {
+#if defined(DNNL_DEV_MODE)
+        return get_verbose(verbose_t::debuginfo) >= static_cast<int>(level);
+#else
+        return false;
+#endif
+    }
+
+    log_level_t get_level() const { return level; }
+
+    operator bool() const { return value; }
+
+    template <typename T>
+    logger_t &operator<<(const T &obj) {
+        //using dnnl::impl::gpu::intel::jit::operator<<;
+        maybe_print_header();
+        std::ostringstream oss;
+        print_helper_t<T>::call(oss, obj);
+        auto lines = gpu_utils::split(oss.str(), "\n");
+        for (size_t i = 0; i < lines.size(); i++) {
+            if (ends_with_endl_) out_ << "  ";
+            out_ << lines[i];
+            ends_with_endl_ = false;
+            if (i != lines.size() - 1) {
+                out_ << "\n";
```

Comment:
I think we should emit a prefix for every verbose line. Currenlty the log looks like:
```
[TRACE][        trace.cpp:39] === After inject_slm_reorder
  alloc slm[0] (mem_usage: 0)
  local_id2.u16 = (nil)
```

But I would expect it to be more like:
```
[TRACE][        trace.cpp:39] === After inject_slm_reorder
[TRACE][        trace.cpp:39]   alloc slm[0] (mem_usage: 0)
[TRACE][        trace.cpp:39]   local_id2.u16 = (nil)
```
or maybe 
```
[TRACE][        trace.cpp:39] === After inject_slm_reorder
[TRACE]                         alloc slm[0] (mem_usage: 0)
[TRACE]                         local_id2.u16 = (nil)
```

Reply:
This is a good question - I thought about that. The issue I see with you suggestion is that for the most part such prefixes just eat space without really helping much, e.g.:


![image](https://github.com/user-attachments/assets/573601cc-eaba-4968-853d-f2412d67205b)

Here we have a prefix with ~50 chars which in useless in practice so why don't we just put a short indent for continuation lines and keep logs in the more convenient/readable form?

Reply:
> Here we have a prefix with ~50 chars which in useless in practice so why don't we just put a short indent for continuation lines and keep logs in the more convenient/readable form?

I agree, I think including the filename in the prefix is a bad idea. What if we used different logging structures based on supplied data? I am thinking something like:

```
[prefix] [path/to/filename:line] <short message format>

[prefix] [path/to/filename:line]
[prefix]     <long message format>

[prefix] [path/to/filename:line]
[prefix]     <multiline>
[prefix]     <message>
[prefix]     <format>
```
This way we can keep the prefix short, while still logging relevant file information at the cost of a small amount of buffering and logic.

Reply:
Yeah, I think this should work better, thanks for the idea though I would suggest to distinguish between single/multi line format only for simplicity.
I'll update the PR. Let's try to use it for a while, and if we see it's not good enough we can update it later.